### PR TITLE
Initial commit for supporting mass provision

### PR DESCRIPTION
**Pull request by ramu-phanimukla**
This is initial commit. More test cases will be added. Before I move far, need quick feedback

### CODE REVIEW

File: autoprotocol/protocol.py [hunk 1]
```diff
@@ -5080,51 +5081,62 @@ def provision(self, resource_id, dests, volumes):
...
-                raise TypeError("Volume must be a string or Unit.")
+            amounts = [Unit(v) for v in amounts]
+        unique_measure_modes = set()
+        for amount in amounts:
+            if not isinstance(amount, (str, Unit)):
+                raise TypeError("Volume/Mass must be a string or Unit.")
+            if amount.unit in VOLUMETRIC_UNITS:
+                unique_measure_modes.add("volume")
+            elif amount.unit in MASS_UNITS:
+                unique_measure_modes.add("mass")
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5102]: hmm we should probably explicitly fail here instead of silently fail? (i.e. else case)

Comment (yangchoo) [inline autoprotocol/protocol.py:5102]: If I understand the intention behind this change, the intention here is to ensure that we generate units which are part of the Autoprotocol-spec.

If that's the intention, I think it'll make more sense to check this during the final serialization portion, i.e. in https://github.com/autoprotocol/autoprotocol-python/blob/master/autoprotocol/protocol.py#L5810 or perhaps upstream in the `as_dict` call.

That way, it'll be more applied to all Unit references, not just provision. 

If we were to do that, something else we need to think about is we could potentially automatically convert it to the smallest/largest supported unit. For example, someone specifying "1000:nanoliter" is technically valid when converted to "1:microliter", and they're equivalent.

File: autoprotocol/protocol.py [hunk 2]
```diff
@@ -5080,51 +5081,62 @@ def provision(self, resource_id, dests, volumes):
...
+                            self.provision(resource_id, d, diff)
+                        )
+                        diff -= diff
+                    continue
+
+                if d.volume:
+                    d.volume += amount
+                else:
+                    d.set_volume(amount)
+            self._remove_cover(d.container, "provision")
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5138]: remove cover call should happen before we start provisioning

File: autoprotocol/protocol.py [hunk 3]
```diff
@@ -5080,51 +5081,62 @@ def provision(self, resource_id, dests, volumes):
...
             xfer["well"] = d
-            xfer["volume"] = v
+            xfer[measurement_mode] = amount
 
-            if d.volume:
-                d.volume += v
-            else:
-                d.set_volume(v)
-            dest_group.append(xfer)
+            if measurement_mode == "volume":
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5115]: We briefly talked about the possibility of doing the max_vol check by converting with density... Probably more of a nice-to-have

File: test/provision_test.py [hunk 4]
```diff
@@ -183,3 +183,20 @@ def test_for_multiple_dispenses_of_resource_in_containers_larger_than_900ml(self
...
+        )
+        p.provision("rs17gmh5wafm5p", w1, "50:ug")
+        actual_instruction_as_json = json.dumps(
+            p.as_dict()["instructions"], indent=2, sort_keys=True
+        )
+        expected_instruction_as_json = TestUtils.read_json_file(
+            "provision_for_mass.json"
+        )
+
+        assert expected_instruction_as_json == actual_instruction_as_json
```
Comment (yangchoo) [inline test/provision_test.py:202]: lets try to add more failure cases

File: autoprotocol/constants.py [hunk 5]
```diff
@@ -35,3 +35,9 @@
     (0.6243159222654036, -0.14191423819668741),
     (0.563207308766221, 0.5152506858187523),
 ]
+
+MEASUREMENT_MODES = ["mass", "volume"]
+
+VOLUMETRIC_UNITS = ["liter", "milliliter", "microliter"]
+
+MASS_UNITS = ["gram", "milligram", "microgram"]
```
Comment (yangchoo) [inline autoprotocol/constants.py:43]: See below, something to consider is to abstract this out more generically

File: autoprotocol/protocol.py (hunk 1 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5102]: Actually below code will catch(I know message is not capturing that).
```
        if len(unique_measure_modes) != 1:
            raise ValueError("Received amounts with more than one measurement mode")
```

Anyhow I added else block.

File: autoprotocol/protocol.py (hunk 2 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5138]: Good catch.

File: autoprotocol/protocol.py (hunk 3 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5115]: Since there is no consensus yet on this, I will wait on product owner to conform on the requirement. May be I will reach out to apps team try to get their input. But I prefer to have PO take final call and put it in the issue.

File: test/provision_test.py (hunk 4 above)
Comment (ramu-phanimukla) [reply test/provision_test.py:202]: My commit message was :

> This is initial commit. More test cases will be added. Before I move far, need quick feedback

File: test/data/provision_for_mass.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
+[
+  {
+    "measurement_mode": "mass",
```
Comment (joshhacksthings) [inline test/data/provision_for_mass.json:3]: just curious why `"measurement_mode"` is sorted to the top here when set to `"mass"`, but not for `"volume"`, when `json.dumps(sort_key=True)` sorts by keys.

File: autoprotocol/constants.py (hunk 5 above)
Comment (joshhacksthings) [inline autoprotocol/constants.py:43]: there are more volumetric and mass units than these. is it possible not to hardcode here?

File: autoprotocol/protocol.py [hunk 7]
```diff
@@ -5058,19 +5058,20 @@ def provision(self, resource_id, dests, volumes):
         TypeError
             If resource_id is not a string.
         RuntimeError
-            If length of the list of volumes specified does not match the number
+            If length of the list of volumes/masses specified does not match the number
             of destination wells specified.
         TypeError
-            If volume is not specified as a string or Unit (or a list of either)
+            If volume/mass is not specified as a string or Unit (or a list of either)
```
[the remark this answers is not in the document]
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5064]: Makes sense. Fixed it through-out this method.

File: test/data/provision_for_mass.json (hunk 6 above)
Comment (ramu-phanimukla) [reply test/data/provision_for_mass.json:3]: In json file can have the keys in any order. But at load time(with the option of `sort_keys=True`), it will sort the keys. 

The main reason for specifying the option of `sort_keys=True` is that, when developer defines json(with keys in any order) and there is no library to support comparison of logical structure and data of expected v/s actual json, I am relying on string comparison(for which keys need to be sorted)

File: autoprotocol/constants.py (hunk 5 above)
Comment (ramu-phanimukla) [reply autoprotocol/constants.py:43]: @joshhacksthings I know there are more volumetric and mass units. As an implementer AP specification, I read http://autoprotocol.org/specification/#units and assumed that, in the scientific world only mentioned units makes sense and hence scoped down to this list.

As per the ASC https://github.com/strateos/autoprotocol_standard_changes/blob/master/Proposed/asc052.md, in order to populate `measurement_mode` attribute, I need to refer to these constants. If you can point me to some library which can return some indication of measurement made based on unit, I will get rid of these constants. Till then, we may have to retain(I will add other possible units for volume and mass). Let me know.

**Comment by ramu-phanimukla** [comment]
> I would prefer if we separated out the Unit check from this diff... I think its definitely something worth considering, but it'll be something that'll make more sense if its consistently applied.
Can you please elaborate bit more on how you are envisioning this?

> We should probably also decide on if we want to do the mass check via density conversion.
Based on the comment https://strateos.atlassian.net/browse/SEA-240?focusedCommentId=16025, from product we are not performing any checks now.

File: autoprotocol/protocol.py (hunk 3 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5115]: Based on [comment](https://strateos.atlassian.net/browse/SEA-240?focusedCommentId=16025)  from product, we are not adding any checks or logic as part of this US implementation

**Comment by ramu-phanimukla** [comment]
@yangchoo & @joshhacksthings 

I removed static definitions from constants file and using string functions to decide measurement mode. Have a look at the changes. The main intention of adding those constants was trying to adhere to whatever was listed in [AP specification](http://autoprotocol.org/specification/#units).

File: autoprotocol/constants.py [hunk 8]
```diff
@@ -35,3 +35,5 @@
     (0.6243159222654036, -0.14191423819668741),
     (0.563207308766221, 0.5152506858187523),
 ]
+
+MEASUREMENT_MODES = ["mass", "volume"]
```
Comment (yangchoo) [inline autoprotocol/constants.py:39]: nit: Let's rename this to `PROVISION_MEASUREMENT_MODES`, since there are other measurement modes (e.g. Spectrophotometry)

File: autoprotocol/protocol.py [hunk 9]
```diff
@@ -5047,9 +5047,9 @@ def provision(self, resource_id, dests, volumes):
         dests : Well or WellGroup or list(Well)
           Destination(s) for specified resource.
         volumes : str or Unit or list(str) or list(Unit)
-          Volume(s) to transfer of the resource to each destination well.  If
-          one volume of specified, each destination well recieve that volume of
-          the resource.  If destinations should recieve different volumes, each
+          Volume(s) or Mass(es) to transfer of the resource to each destination well.  If
+          one volume/mass is specified, each destination well receive that volume/mass of
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5051]: nit: think you missed the `/` here

File: autoprotocol/protocol.py [hunk 10]
```diff
@@ -5047,9 +5047,9 @@ def provision(self, resource_id, dests, volumes):
         dests : Well or WellGroup or list(Well)
           Destination(s) for specified resource.
         volumes : str or Unit or list(str) or list(Unit)
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5049]: I understand there was the comment with @joshhacksthings on this... to use `volumes` rather than `amounts` for backwards compat. 

However, having `volumes` as the docstring/argument name seems pretty unintuitive. One possible workaround to support backwards compat would be to add volumes as a kwarg (keyword arugment).
I.e. `provision(self, resource_id, dests, amounts, volumes=None)`

Hmm.. for now we can add a comment that it's for backwards compat/will be deprecated. Forward thinking, we could look into something similar to https://pypi.org/project/Deprecated/ as well. That would be nice for marking these.

What do you think, @joshhacksthings ?

File: autoprotocol/protocol.py [hunk 11]
```diff
@@ -5135,10 +5141,30 @@ def provision(self, resource_id, dests, volumes):
...
+                    )
                 )
         return provision_instructions_to_return
 
+    def _identify_measurement_mode(self, volumes):
+        unique_measure_modes = set()
+        for amount in volumes:
+            if not isinstance(amount, (str, Unit)):
+                raise TypeError(f"Provided amount {amount} is not supported.")
+            if amount.unit.endswith("liter"):
```
[the remark this answers is not in the document]
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py]: @yangchoo & @joshhacksthings 
I prefer using below as it will be more generic:
```
# For mass
amount.dimensionality == Unit(1, "gram").dimensionality 
# For volume
amount.dimensionality == Unit(1, "liter").dimensionality
```
Modified the code accordingly

File: autoprotocol/protocol.py (hunk 9 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py]: Done

File: autoprotocol/constants.py (hunk 8 above)
Comment (ramu-phanimukla) [reply autoprotocol/constants.py]: Renamed.

File: autoprotocol/constants.py [hunk 12]
```diff
@@ -35,3 +35,5 @@
     (0.6243159222654036, -0.14191423819668741),
     (0.563207308766221, 0.5152506858187523),
 ]
+
+PROVISION_MEASUREMENT_MODES = ["mass", "volume"]
```
Comment (joshhacksthings) [inline autoprotocol/constants.py:39]: nit: could we move this into the `Provision` Instruction? or the protocol method?

File: autoprotocol/instruction.py [hunk 13]
```diff
@@ -9,6 +9,7 @@
 
 # pragma pylint: disable=too-few-public-methods, redefined-builtin
 from .builders import *  # pylint: disable=unused-wildcard-import
+from .constants import *  # pylint: disable=unused-wildcard-import
```
Comment (joshhacksthings) [inline autoprotocol/instruction.py:12]: nit: if we know for sure we're only importing one object (`PROVISION_MEASUREMENT_MODES`) I think it would be worth explicitly importing it, and avoiding a wildcard import

File: autoprotocol/instruction.py [hunk 14]
```diff
@@ -1385,9 +1388,19 @@ class Provision(Instruction):
 
     """
 
-    def __init__(self, resource_id, dests):
+    def __init__(self, resource_id, dests, measurement_mode="volume"):
+        if measurement_mode not in PROVISION_MEASUREMENT_MODES:
+            raise RuntimeError(
```
Comment (joshhacksthings) [inline autoprotocol/instruction.py:1393]: add `RuntimeError` to the "Raises" section of this function's docstring

File: docs/changelog.rst [hunk 15]
```diff
@@ -2,6 +2,8 @@
 Changelog
 =========
 
+* :support:`255` Add support for provisioning mass based resources
```
Comment (joshhacksthings) [inline docs/changelog.rst:5]: nit: I'd argue this could be a `feature`

File: autoprotocol/protocol.py [hunk 16]
```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
...
+                    self._append_and_return(
+                        Provision(resource_id, dest_group, measurement_mode)
+                    )
                 )
         return provision_instructions_to_return
 
+    def _identify_measurement_mode(self, volumes):
+        unique_measure_modes = set()
+        for amount in volumes:
+            print(amount.dimensionality)
```
Comment (joshhacksthings) [inline autoprotocol/protocol.py:5153]: remove `print`

File: autoprotocol/protocol.py [hunk 17]
```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
...
             else:
                 provision_instructions_to_return.append(
-                    self._append_and_return(Provision(resource_id, dest_group))
+                    self._append_and_return(
+                        Provision(resource_id, dest_group, measurement_mode)
+                    )
                 )
         return provision_instructions_to_return
 
+    def _identify_measurement_mode(self, volumes):
```
Comment (joshhacksthings) [inline autoprotocol/protocol.py:5150]: nit: rename `volumes` argument to something more descriptive. It looks like we're expecting a list of `Unit`s. I'd be fine with `amounts` or something like that

File: autoprotocol/protocol.py [hunk 18]
```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
...
+        unique_measure_modes = set()
+        for amount in volumes:
+            print(amount.dimensionality)
+            if not isinstance(amount, (str, Unit)):
```
Comment (joshhacksthings) [inline autoprotocol/protocol.py:5154]: this check is a bit unwieldy. 

First, we should not be allowing `amount` to be of type `str` as the subsequent steps call an attribute that only `Unit` objects have.

Second, we could use something like `amount = parse_unit(amount)` to sanely parse the unit provided. It will guarantee `amount` is type `Unit`

File: autoprotocol/protocol.py [hunk 19]
```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
...
+            if amount.dimensionality == Unit(1, "liter").dimensionality:
+                unique_measure_modes.add("volume")
+            elif amount.dimensionality == Unit(1, "gram").dimensionality:
+                unique_measure_modes.add("mass")
+            else:
+                raise ValueError(
+                    f"Provisioning of resources with measurement unit of {amount.unit} is not supported."
+                )
+        if len(unique_measure_modes) != 1:
+            raise ValueError("Received amounts with more than one measurement mode")
```
Comment (joshhacksthings) [inline autoprotocol/protocol.py:5165]: Up until this line, this method has been instruction agnostic. ie, you could use `_identify_measurement_mode` to decipher a measurement mode from any list of `Unit`s provided.

Starting with this `else` block, we raise a `ValueError` based on the `provision` method's support of a type of unit. I think we should make this method clearer, maybe renaming to `_identify_provision_mode`, so it's clear this is a provision-specific function

File: autoprotocol/protocol.py [hunk 20]
```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
...
+        measurement_mode = unique_measure_modes.pop()
```
Comment (joshhacksthings) [inline autoprotocol/protocol.py:5166]: nit: would this be a good place to check that the single item in `unique_measure_mode` is also in `PROVISION_MEASUREMENT_MODES`?

File: test/data/provision_for_mass.json (hunk 6 above)
Comment (joshhacksthings) [reply test/data/provision_for_mass.json:3]: Got it, makes total sense. Since we can have these keys in any order, can we move `measurement_mode` to follow the line `"op": "provision",` just to harmonize with other provision json files?

File: test/protocol_test.py [hunk 21]
```diff
@@ -2738,9 +2738,9 @@ def test_convert_provision(self):
         p1 = Protocol()
         c1 = p1.ref("c1", id=None, cont_type="96-pcr", discard=True)
         p1.incubate(c1, where="ambient", duration="1:hour", uncovered=True)
-        p1.provision("rs18s8x4qbsvjz", c1.well(0), volumes="10:microliter")
+        p1.provision("rs18s8x4qbsvjz", c1.well(0), "10:microliter")
         p1.incubate(c1, where="ambient", duration="1:hour", uncovered=True)
-        p1.provision("rs18s8x4qbsvjz", c1.well(0), volumes="10:microliter")
+        p1.provision("rs18s8x4qbsvjz", c1.well(0), "10:microliter")
```
Comment (joshhacksthings) [inline test/protocol_test.py:2743]: let's keep the `volumes` kwarg, just to ensure backwards compat

File: test/provision_test.py [hunk 22]
```diff
@@ -174,7 +174,7 @@ def test_for_multiple_dispenses_of_resource_in_containers_larger_than_900ml(self
             .well(0)
             .set_volume("2:microliter")
         )
-        p.provision("rs17gmh5wafm5p", w1, "1500:microliter")
+        p.provision("rs17gmh5wafm5p", w1, volumes="1500:microliter")
```
Comment (joshhacksthings) [inline test/provision_test.py:177]: same as above, let's keep the volumes kwarg, just to ensure backwards compat

File: test/provision_test.py [hunk 23]
```diff
@@ -183,3 +183,45 @@ def test_for_multiple_dispenses_of_resource_in_containers_larger_than_900ml(self
             "split_provisions_by_volume.json"
         )
         assert expected_instruction_as_json == actual_instruction_as_json
+
+    def test_provision_well_with_mass(self):
+        p = Protocol()
```
Comment (joshhacksthings) [inline test/provision_test.py:188]: I'd strongly recommend writing these `Protocol` and `Well` objects as pytest fixtures, just to cutdown on rewriting boilerplate

a good example can be found in [TestTransferVolume](https://github.com/autoprotocol/autoprotocol-python/blob/3d09f1271e3695a260b0ce4e85a8445a3ef38b8e/test/protocol_test.py#L3258)

File: autoprotocol/constants.py (hunk 12 above)
Comment (ramu-phanimukla) [reply autoprotocol/constants.py:39]: As they are really serving as constants and we have separate file for that, I feel it is the right place.

File: test/provision_test.py (hunk 23 above)
Comment (ramu-phanimukla) [reply test/provision_test.py]: Updated `provision_test.py` tests to use `pytest fixtures`

File: test/provision_test.py (hunk 22 above)
Comment (ramu-phanimukla) [reply test/provision_test.py]: Done

File: test/protocol_test.py (hunk 21 above)
Comment (ramu-phanimukla) [reply test/protocol_test.py]: Done

File: autoprotocol/protocol.py (hunk 20 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5184]: I did thought about it, but it will be redundant as we already have that check in the `Provision` constructor.

File: docs/changelog.rst (hunk 15 above)
Comment (ramu-phanimukla) [reply docs/changelog.rst]: Makes sense. Modified to:
```
* :feature:`255` Add support for provisioning of resources by mass
```

File: autoprotocol/instruction.py (hunk 14 above)
Comment (ramu-phanimukla) [reply autoprotocol/instruction.py:1394]: Updated `Raises` section

File: autoprotocol/instruction.py (hunk 13 above)
Comment (ramu-phanimukla) [reply autoprotocol/instruction.py]: Followed previous import statement. Anyhow fixed it.

File: autoprotocol/instruction.py [hunk 24]
```diff
@@ -1385,9 +1388,17 @@ class Provision(Instruction):
 
     """
 
-    def __init__(self, resource_id, dests):
+    def __init__(self, resource_id, dests, measurement_mode="volume"):
+        if measurement_mode not in MEASUREMENT_MODES:
+            raise RuntimeError(f"{measurement_mode} is not a valid measurement mode")
```
[the remark this answers is not in the document]
Comment (ramu-phanimukla) [reply autoprotocol/instruction.py]: Added `If the measurement mode is not supported.` to "Raises" section of the docstring.

**Comment by ramu-phanimukla** [comment]
@joshhacksthings & @yangchoo All comments addressed

File: test/data/provision_for_mass.json (hunk 6 above)
Comment (ramu-phanimukla) [reply test/data/provision_for_mass.json]: Ensured same order of keys across all files

File: autoprotocol/protocol.py [hunk 25]
```diff
@@ -5034,7 +5034,7 @@ def image_plate(self, ref, mode, dataref):
         """
         return self._append_and_return(ImagePlate(ref, mode, dataref))
 
-    def provision(self, resource_id, dests, volumes):
+    def provision(self, resource_id, dests, volumes=None, amounts=None):
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5037]: Would prefer `def provision(self, resource_id, dests, amounts, volumes=None)` instead. This retains backwards compat support while allowing the user to provide mass as the third argument. I.e. `provision(..., ..., Unit("1:gram"))`

Comment (yangchoo) [inline autoprotocol/protocol.py:5037]: would prefer to swap `volumes` and `amounts` positions, so it'll allow someone to specify mass as the third positional arg

File: autoprotocol/protocol.py (hunk 16 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5171]: Done

File: autoprotocol/protocol.py (hunk 19 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5183]: Renamed to `_identify_provision_mode`

File: autoprotocol/protocol.py (hunk 18 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5172]: I see below code is redundant:
```
            if not isinstance(amount, (str, Unit)):
                raise TypeError(f"Provided amount {amount} is not supported.")
```
because it is handled by below code which is before `_identify_measurement_mode ` call:
```
        if not isinstance(provision_amounts, list):
            provision_amounts = [Unit(provision_amounts)] * len(dests)
        else:
            if len(provision_amounts) != len(dests):
                raise RuntimeError(
                    "To provision a resource into multiple "
                    "destinations with multiple volumes or masses, the  "
                    "list of volumes or masses must correspond with the "
                    "destinations in length and in order."
                )
            provision_amounts = [Unit(v) for v in provision_amounts]
```

File: autoprotocol/protocol.py (hunk 17 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py]: Renamed `provision_amounts`

File: autoprotocol/protocol.py (hunk 25 above)
Comment (yangchoo) [reply autoprotocol/protocol.py]: done, thanks!

File: autoprotocol/protocol.py [hunk 26]
```diff
@@ -5046,85 +5046,109 @@ def provision(self, resource_id, dests, volumes):
           Resource ID from catalog.
         dests : Well or WellGroup or list(Well)
           Destination(s) for specified resource.
+        amounts : str or Unit or list(str) or list(Unit)
+          Mass(es) to transfer of the resource to each destination well.  If
+          one mass is specified, each destination well receive that mass of
+          the resource.  If destinations should receive different mass, each
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5052]: Should this be `Volumes or Masses`?

File: autoprotocol/protocol.py (hunk 18 above)
Comment (yangchoo) [reply autoprotocol/protocol.py:5171]: Ah I see. Its already been cast by the time it reaches here. 

In that case, I think its fine to remove the `str` type-check. I.e. just ` if not isinstance(amount, Unit)`

File: autoprotocol/protocol.py (hunk 20 above)
Comment (joshhacksthings) [reply autoprotocol/protocol.py:5183]: Yeah, I was thinking that the check here would be in place of that one. I'll leave it up to you

File: autoprotocol/protocol.py [hunk 27]
```diff
@@ -5046,85 +5046,109 @@ def provision(self, resource_id, dests, volumes):
...
-                    "list of volumes must correspond with the "
+                    "destinations with multiple volumes or masses, the  "
+                    "list of volumes or masses must correspond with the "
                     "destinations in length and in order."
                 )
-            volumes = [Unit(v) for v in volumes]
-        for v in volumes:
-            if not isinstance(v, (str, Unit)):
-                raise TypeError("Volume must be a string or Unit.")
+            provision_amounts = [Unit(v) for v in provision_amounts]
```
Comment (yangchoo) [inline autoprotocol/protocol.py:5116]: missed this from the previous review, but fwiw, you can do `[parse_unit(v, ["liter", "gram") for v in provision_amounts]` instead of `[Unit(v) for v in provision_amounts]`

That way, there'll be additional type checks earlier on. Its not expensive and wraps it with nicer error messages

File: autoprotocol/instruction.py [hunk 28]
```diff
@@ -9,6 +9,7 @@
 
 # pragma pylint: disable=too-few-public-methods, redefined-builtin
 from .builders import *  # pylint: disable=unused-wildcard-import
+from .constants import PROVISION_MEASUREMENT_MODES
```
Comment (joshhacksthings) [inline autoprotocol/instruction.py:12]: pasting conversation from DMs here:

Ramu: 
I was following the previous import statement and if I make change, we might have inconsistency. What do you think?
me:
The import above this (from `.builders` is fine using a wildcard import because the code in `protocol.instruction` depends on all 11 classes defined in `.builders`.
We only need a single object, `PROVISION_MEASUREMENT_MODES` from `.constants`.

File: autoprotocol/protocol.py [hunk 29]
```diff
@@ -5046,85 +5046,109 @@ def provision(self, resource_id, dests, volumes):
...
         volumes : str or Unit or list(str) or list(Unit)
-          Volume(s) to transfer of the resource to each destination well.  If
-          one volume of specified, each destination well recieve that volume of
-          the resource.  If destinations should recieve different volumes, each
+          Volume to transfer of the resource to each destination well.  If
+          one volume is specified, each destination well receive that volume of the resource.
+          If destinations should receive different volumes, each
           one should be specified explicitly in a list matching the order of the
           specified destinations.
+          Note: If value is provided for this field, value for 'amounts' must be skipped
```
Comment (joshhacksthings) [inline autoprotocol/protocol.py:5062]: nit: I'd rephrase to something like `volumes` and `amounts` arguments are mutually exclusive. Only one is required.

File: autoprotocol/protocol.py [hunk 30]
```diff
@@ -5046,85 +5046,109 @@ def provision(self, resource_id, dests, volumes):
...
         dests = WellGroup(dests)
+
         if not isinstance(resource_id, str):
             raise TypeError("Resource ID must be a string.")
-        if not isinstance(volumes, list):
-            volumes = [Unit(volumes)] * len(dests)
+
+        if (volumes is None and amounts is None) or (volumes and amounts):
+            raise ValueError(
+                "Either volumes or amounts should have value/s, but not both."
```
Comment (joshhacksthings) [inline autoprotocol/protocol.py:5098]: nit: change "value/s" to "value(s)"

File: autoprotocol/protocol.py (hunk 18 above)
Comment (joshhacksthings) [reply autoprotocol/protocol.py:5171]: Yep

File: test/protocol_test.py [hunk 31]
```diff
@@ -2740,7 +2740,7 @@ def test_convert_provision(self):
         p1.incubate(c1, where="ambient", duration="1:hour", uncovered=True)
         p1.provision("rs18s8x4qbsvjz", c1.well(0), volumes="10:microliter")
         p1.incubate(c1, where="ambient", duration="1:hour", uncovered=True)
-        p1.provision("rs18s8x4qbsvjz", c1.well(0), volumes="10:microliter")
+        p1.provision("rs18s8x4qbsvjz", c1.well(0), "10:microliter")
```
Comment (joshhacksthings) [inline test/protocol_test.py:2743]: nit: I'd keep the `volumes=` key word here since the behavior should remain unchanged

File: test/provision_test.py [hunk 32]
```diff
@@ -88,21 +80,15 @@ def test_with_consecutive_repeated_wells(self):
         assert expected_instruction_as_json == actual_instruction_as_json
 
     def test_with_repeated_wells_but_discontinious(self):
```
Comment (joshhacksthings) [inline test/provision_test.py:82]: nit: correct typo `discontinious` to `discontinuous` (since we're changing this method I feel it's within scope)

File: test/provision_test.py [hunk 33]
```diff
@@ -168,18 +154,46 @@ def test_provision_with_sealed_container(self):
...
-            p.as_dict()["instructions"], indent=2, sort_keys=True
+            self.p.as_dict()["instructions"], indent=2, sort_keys=True
         )
         expected_instruction_as_json = TestUtils.read_json_file(
             "split_provisions_by_volume.json"
         )
         assert expected_instruction_as_json == actual_instruction_as_json
+
+    def test_provision_well_with_mass(self):
+        self.p.provision("rs17gmh5wafm5p", self.w1, amounts="50:ug")
```
Comment (joshhacksthings) [inline test/provision_test.py:174]: nit: we're missing a test case that correctly provisions more than one mass

File: autoprotocol/protocol.py (hunk 20 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5183]: If someone is trying to create provision object like :
```
provision = Provision(resource_id, dest_group, measurement_mode)
```
the check in the constructor will prevent invalid measurement_mode being set in the object

File: autoprotocol/protocol.py (hunk 26 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5052]: Corrected the documentation

File: autoprotocol/protocol.py (hunk 27 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5116]: Fixed

File: autoprotocol/instruction.py [hunk 34]
```diff
@@ -1380,14 +1383,25 @@ class Provision(Instruction):
     RuntimeError
       If length of the list of volumes specified does not match the number of
       destination wells specified.
+      If the measurement mode is not supported.
```
[the remark this answers is not in the document]
Comment (ramu-phanimukla) [reply autoprotocol/instruction.py:1386]: Fixed

File: autoprotocol/protocol.py (hunk 26 above)
Comment (yangchoo) [reply autoprotocol/protocol.py:5052]: thanks!

File: autoprotocol/protocol.py (hunk 29 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5062]: Fixed

File: autoprotocol/protocol.py (hunk 30 above)
Comment (ramu-phanimukla) [reply autoprotocol/protocol.py:5098]: Fixed

File: test/protocol_test.py (hunk 31 above)
Comment (ramu-phanimukla) [reply test/protocol_test.py:2743]: Fixed

File: test/provision_test.py (hunk 32 above)
Comment (ramu-phanimukla) [reply test/provision_test.py:82]: Fixed

File: test/provision_test.py (hunk 33 above)
Comment (ramu-phanimukla) [reply test/provision_test.py:174]: Added explicit test case `test_provision_multiple_wells_with_diff_masses`

File: docs/changelog.rst [hunk 35]
```diff
@@ -2,6 +2,8 @@
 Changelog
 =========
 
+* :feature:`255` Add support for provisioning of resources by mass
```
Comment (yangchoo) [inline docs/changelog.rst:5]: this should be 254

File: test/protocol_test.py (hunk 31 above)
Comment (joshhacksthings) [reply test/protocol_test.py:2743]: still missing `volumes`

original line is:
`p1.provision("rs18s8x4qbsvjz", c1.well(0), volumes="10:microliter")`
